src/plugins/custom/file-handling.ts

```diff
@@ -0,0 +1,76 @@
+import fp from 'fastify-plugin'
+import { BusboyFileStream } from '@fastify/busboy'
+import { FastifyInstance } from 'fastify'
+import path from 'path'
+import { pipeline } from 'node:stream/promises'
+import fs from 'node:fs'
+
+declare module 'fastify' {
+  export interface FastifyInstance {
+    fileHandler: {
+      upload: (fileDir: FileDir, fileName: string, file: BusboyFileStream) => Promise<void>
+      remove: (fileDir: FileDir, fileName: string) => Promise<void>
+    }
+  }
+
+}
+
+interface FileDirConfig {
+  UPLOAD_TASKS_DIRNAME: string
+}
+
+function buildFilePath (fileDir: FileDir, filename: string, fastify: FastifyInstance) {
+  return path.join(
+    import.meta.dirname,
+    '../../../',
+    fastify.config.UPLOAD_DIRNAME,
+    fastify.config[fileDirMap[fileDir]],
+    filename
+  )
+}
+
+type FileDir = 'tasks'
+
+type FileDirConfigPath<T> = {
+  [K in keyof T]: K extends `UPLOAD_${string}_DIRNAME` ? K : never
+}[keyof T]
+
+type FileDirMap = {
+  [key in FileDir]: FileDirConfigPath<FileDirConfig>
+}
+
+const fileDirMap: FileDirMap = {
+  tasks: 'UPLOAD_TASKS_DIRNAME'
+}
+
+async function upload (this: FastifyInstance, fileDir: FileDir, fileName: string, file: BusboyFileStream) {
+  const filePath = buildFilePath(fileDir, fileName, this)
+
+  await pipeline(file, fs.createWriteStream(filePath))
+}
+
+async function remove (this: FastifyInstance, fileDir: FileDir, fileName: string) {
+  const filePath = buildFilePath(fileDir, fileName, this)
+  try {
+    await fs.promises.unlink(filePath)
+  } catch (err) {
+    if (isErrnoException(err) && err.code === 'ENOENT') {
+      // A file could have been deleted by an external actor, e.g. system administrator.
+      // We log the error to keep a record of the failure, but consider that the operation was successful.
+      this.log.warn(`File path '${fileName}' not found`)
+    } else {
+      throw err
+    }
+  }
+}
+
+function isErrnoException (error: unknown): error is NodeJS.ErrnoException {
+  return error instanceof Error && 'code' in error
+}
+
+export default fp(async (fastify) => {
+  fastify.decorate('fileHandler', {
+    upload: upload.bind(fastify),
+    remove: remove.bind(fastify)
```

Comment:
This is not needed. 

Fastify instance (`this`) will be automatically passed if we use function declaration with `function` keyword (doesn't work with arrow functions).

Reply:
I have to bind it manually, otherwise the instance passed is incorrect.. (empty object) 
I think that it is because of the way i use the `decorate` function, I passe to it an objects and not a function, so the functions inside of the object doesn't get the right instance.

Reply:
> I passe to it an objects and not a function

Ha, ok I reviewed too quickly.

You can put all the functions you use directly in the registration callback so you get access to the fastify instance.
```ts
export default fp(async (fastify) => {
// do all the logic here
```

Reply:
It's works, but i can't do this (upload and remove not anymore in scope)
```typescript
declare module 'fastify' {
  export interface FastifyInstance {
    fileHandler: {
      upload: typeof upload
      remove: typeof remove
    }
  }
}
```